units/stats_task_end.go

```diff
@@ -116,25 +116,23 @@ func (j *collectTaskEndDataJob) Run(ctx context.Context) {
 	ctx, cancel = context.WithCancel(ctx)
 	defer cancel()
 
-	mgrOpts := cloud.ManagerOpts{
-		Provider: j.host.Provider,
-		Region:   cloud.GetRegion(j.host.Distro),
-	}
-	manager, err := cloud.GetManager(ctx, j.env, mgrOpts)
-	if err != nil {
-		j.AddError(err)
-		grip.Error(message.WrapErrorf(err, "Error loading provider for host %s cost calculation", j.task.HostId))
-	} else {
-		if calc, ok := manager.(cloud.CostCalculator); ok {
-			cost, err = calc.CostForDuration(ctx, j.host, j.task.StartTime, j.task.FinishTime)
-			if err != nil {
-				j.AddError(err)
-			} else {
-				if err = j.task.SetCost(cost); err != nil {
-					j.AddError(err)
-				}
-				if err = j.host.IncCost(cost); err != nil {
+	mgrOpts, err := cloud.GetManagerOptions(j.host.Distro)
+	j.AddError(errors.Wrapf(err, "can't get ManagerOpts for '%s'", j.host.Id))
+	if !j.HasErrors() {
+		manager, err := cloud.GetManager(ctx, j.env, mgrOpts)
+		j.AddError(errors.Wrapf(err, "Error loading provider for host %s cost calculation", j.task.HostId))
+		if err == nil {
+			if calc, ok := manager.(cloud.CostCalculator); ok {
+				cost, err = calc.CostForDuration(ctx, j.host, j.task.StartTime, j.task.FinishTime)
+				if err != nil {
 					j.AddError(err)
+				} else {
+					if err = j.task.SetCost(cost); err != nil {
```

Comment:
nit: i know you didn't write this, but you can get rid of these nested ifs by just wrapping the function in j.AddError since nils don't get added.

Reply:
I refactored this into a separate function that should make the logic more straightforward.